### PR TITLE
Tracing abstraction

### PR DESCRIPTION
**Pull request by spencergibb**
The abstraction is based heavily off of http://htrace.incubator.apache.org.  

The terminology is based on google [dapper](http://research.google.com/pubs/pub36356.html).

* __Trace__: The overarching unit of sleuth.  A tree of spans.  Has an id.
* __Span__: Nodes in a trace tree.  Edges represent parent/child relationship.  Spans without a parent are root spans.  Has a name, id, trace id, parent span ids, start and end times.
* __Annotations__: Data associated with a span
* __Sampler__: Determines when to trace

### API

Trace is in interface with DefaultTrace is an implementation.  It's job is to start Spans appropriately.

* `TraceScope startSpan(String name)`
* `TraceScope startSpan(String name, Span parent)`
* `TraceScope startSpan(String name, Sampler<T> s)`

TraceScope is a class that defines behavior for the life of a Span.

* `public TraceScope(ApplicationEventPublisher publisher, Span span, Span savedSpan)`
* `public Span detach()`
* `public void close()`

You'll notice the `ApplicationEventPublisher` in the constructor.  Currently two spring application events are published: SpanStartedEvent and SpanStoppedEvent, which are self explanatory.  If have implemented listeners that for setting the trace id and span id via Slf4j MDC so those values can be shown in logs.  This replicates the 4financeit correlationid that was printed in the logs.  This is also where publishing spans to various pipelines can be implemented: Zipkin, cloudfoundry, jdbc etc...  Because it is event driven, multiple entities can consume spans.

Sampler interface.  Allows customization of when to trace.  A default Sampler is an argument to `DefaultTrace`.

### Sampler

* `boolean next(T info)`

### Instrumentation

Instrumenting code is the next important step in tracing.  There is instrumentation for hystrix, spring scheduling, spring mvc controllers, servlet filter, `RestTemplate` and Zuul.  There are two major ways of instrumenting code, via AOP and via lifecycle hooks (spring mvc, RestTemplate).

@mstine @dsyer @joshlong

@marcingrzejszczak I'd like you to review and ask any questions you might have about how this is different from the implementation you committed.  I believe everything you had is still there.  What you called `Correlation-Id` is now analogous to `Trace-Id`, you just wouldn't care for the `Span-Id`'s if you didn't want to.  It's pluggable now.

### CODE REVIEW

**Comment by marcingrzejszczak** [comment]
I'm on it - 70 files changes... dis gonna be good :P

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/TraceScope.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
+package org.springframework.cloud.sleuth;
+
+import java.io.Closeable;
+
+import lombok.SneakyThrows;
+import lombok.Value;
+import lombok.experimental.NonFinal;
```
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/TraceScope.java]: I'm sorry if I'm out of line and I should only comment on the trace-id related stuff but is it "safe" to use experimental features of Lombok in a library of Spring-Cloud caliber?

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebAutoConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+	@Bean
+	public WebMvcConfigurerAdapter webMvcConfigurerAdapter(
+			TraceHandlerInterceptor handlerInterceptor) {
+		return new TraceWebConfigurer(handlerInterceptor);
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	public FilterRegistrationBean correlationHeaderFilter() {
```
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceWebAutoConfiguration.java]: `traceHeaderFilter`?

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/slf4j/Slf4jSpanStartedListener.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ * @author Spencer Gibb
+ */
+@Slf4j
+public class Slf4jSpanStartedListener implements ApplicationListener<SpanStartedEvent> {
+
+	@Override
+	public void onApplicationEvent(SpanStartedEvent event) {
+		Span span = event.getSpan();
+		//TODO: what log level?
+		log.info("Starting span: {}", span);
```
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/slf4j/Slf4jSpanStartedListener.java]: In our production cases INFO was really a good thing cause we saw where the initalization of the whole business process took place. 

On the other hand we had to introduce filtering if the URIs cause Swagger users flooded our logs with meaningless entries. But after setting the patterns it's really ok right now. So IMO it should be info

File: spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/DefaultTraceTests.java [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+import org.springframework.cloud.sleuth.event.SpanStoppedEvent;
+import org.springframework.cloud.sleuth.sampler.AlwaysSampler;
+import org.springframework.cloud.sleuth.sampler.IsTracingSampler;
+import org.springframework.context.ApplicationEvent;
+import org.springframework.context.ApplicationEventPublisher;
+
+/**
+ * @author Spencer Gibb
+ */
+public class DefaultTraceTests {
```
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/DefaultTraceTests.java]: Why not Spock? :P

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/scheduling/TraceSchedulingAspect.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ * @author Tomasz Nurkewicz, 4financeIT
+ * @author Michal Chmielarz, 4financeIT
+ * @author Marcin Grzejszczak, 4financeIT
+ * @author Spencer Gibb
+ *
+ * @see Trace
+ */
+@Aspect
+public class TraceSchedulingAspect {
```
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/scheduling/TraceSchedulingAspect.java]: That's pretty much exactly the same what we did if I remember it correctly

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/circuitbreaker/TraceCommand.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+import com.netflix.hystrix.HystrixCommandGroupKey;
+import com.netflix.hystrix.HystrixThreadPoolKey;
+import org.springframework.cloud.sleuth.Trace;
+import org.springframework.cloud.sleuth.TraceScope;
+
+/**
+ * Abstraction over {@code HystrixCommand} that wraps command execution with Trace setting
+ *
+ * @see HystrixCommand
+ * @see CorrelationIdUpdater
```
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/circuitbreaker/TraceCommand.java]: I guess this should be removed

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceFilter.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+		this.skipPattern = DEFAULT_SKIP_PATTERN;
+	}
+
+	public TraceFilter(Trace trace, Pattern skipPattern) {
+		this.trace = trace;
+		this.skipPattern = skipPattern;
+	}
+
+	@Override
+	protected void doFilterInternal(HttpServletRequest request,
```
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceFilter.java]: This is the same point of entry that we had - in comes the request and you pass / add "trace-id" headers, right?

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/TraceRestTemplateInterceptor.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+ * Interceptor that verifies whether the trance and span id has been set on the
+ * request and sets them if one or both of them are missing.
+ *
+ * @see org.springframework.web.client.RestTemplate
+ * @see Trace
+ *
+ * @author Marcin Grzejszczak, 4financeIT
+ * @author Spencer Gibb
+ */
+public class TraceRestTemplateInterceptor implements ClientHttpRequestInterceptor {
```
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/TraceRestTemplateInterceptor.java]: This is the same for ensuring that when using `RestTemplate` the requests have the proper headers

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/slf4j/Slf4jSpanStartedListener.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+		MDC.put(Trace.SPAN_ID_NAME, span.getSpanId());
```
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/slf4j/Slf4jSpanStartedListener.java]: And there we have MDC.put with the former 'Correlation-Id' and current 'Trace-id'.

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceFilter.java (hunk 7 above)
Comment (spencergibb) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/TraceFilter.java]: Yes

**Comment by marcingrzejszczak** [comment]
Maybe I'm wrong but we have also this:

https://github.com/4finance/micro-infra-spring/blob/master/micro-infra-spring-base/src/main/groovy/com/ofg/infrastructure/web/correlationid/CorrelationIdAspect.java

Where we try to ensure that if someone is using Servlet 3.0 support with the Callable<> in Controllers that we pass the proper correlationid. I don't see it here - is it deliberate or did you solve it by some other means?

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/slf4j/Slf4jSpanStartedListener.java (hunk 9 above)
Comment (spencergibb) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/slf4j/Slf4jSpanStartedListener.java]: :+1:

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/TraceRestTemplateInterceptor.java (hunk 8 above)
Comment (spencergibb) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/web/client/TraceRestTemplateInterceptor.java]: :+1:

File: spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/DefaultTraceTests.java (hunk 4 above)
Comment (spencergibb) [inline spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/DefaultTraceTests.java]: I'm lazy? ;-)

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/slf4j/Slf4jSpanStartedListener.java (hunk 3 above)
Comment (spencergibb) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/slf4j/Slf4jSpanStartedListener.java]: good to know

File: spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/DefaultTraceTests.java (hunk 4 above)
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/DefaultTraceTests.java]: Oh come on :D You would have done it in Spock in no time with less code :P

**Comment by spencergibb** [comment]
I wondered about Callable in Controllers, right now I trace _ALL_ controllers.  But you're concerned with the thread there right?

**Comment by spencergibb** [comment]
I've replaced references to `correlationId` with references to tracing.

**Comment by marcingrzejszczak** [comment]
Yeah exactly - we've also traced all controllers since we had a filter. However a new thread is spawned and you loose correlationID info. That's why we did the aspect.

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/TraceScope.java (hunk 1 above)
Comment (spencergibb) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/TraceScope.java]: I'm not worried about it.  Lombok doesn't "ship" with spring cloud it's just compile time.

**Comment by spencergibb** [comment]
OK, I'll test that and make sure `Callables` returned from controllers still work.

**Comment by spencergibb** [comment]
That would also be an issue when Observable support https://github.com/spring-cloud/spring-cloud-netflix/pull/246 is in spring cloud netflix.

**Comment by marcingrzejszczak** [comment]
Definitely...

File: spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/DefaultTraceTests.java (hunk 4 above)
Comment (spencergibb) [inline spring-cloud-sleuth-core/src/test/java/org/springframework/cloud/sleuth/DefaultTraceTests.java]: I don't know spock yet, so 'no time' might be a stretch.  I want to learn it though.

File: spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/TraceScope.java (hunk 1 above)
Comment (marcingrzejszczak) [inline spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/TraceScope.java]: :+1:

**Comment by spencergibb** [comment]
Htrace has some [wrappers](https://git-wip-us.apache.org/repos/asf?p=incubator-htrace.git;a=tree;f=htrace-core/src/main/java/org/apache/htrace/wrappers;h=c5e9dc01c82ac3222c40a371661a7f163017b206;hb=HEAD) that could be used there with developer education.

**Comment by marcingrzejszczak** [comment]
Awesome! That would be great!

**Comment by spencergibb** [comment]
@marcingrzejszczak added fix for returning `Callable` from a `Controller`.

**Comment by marcingrzejszczak** [comment]
With Callables and Controllers it seems that everything from our side is there.

BTW @dsyer 2 questions:

1) what next could be merged from our side? 
2) From our perspective it would be cool if you released some 0.0.1 non snapshot version of both Zookeeper and Sleuth. That way we would plug it in into our stuff to see what's missing and how it's actually working. 

WDYT?

**Comment by spencergibb** [comment]
@marcingrzejszczak I'll look at a milestone soon.

**Comment by marcingrzejszczak** [comment]
BTW @spencergibb  - why did you remove all the spock tests? :P you could have reused them

**Pull request merged**